pulser-core/pulser/sequence/_seq_drawer.py

```diff
@@ -161,82 +164,87 @@ def _give_curves_from_samples(
         in_eom_mode = False
         eom_block_n = -1
         # Last eom interval is extended if eom mode not disabled at the end
-        if nb_eom_intervals > 0 and seq.get_duration() == eom_intervals[-1].tf:
+        if (
+            nb_eom_intervals > 0
+            and ch_samples.duration == eom_intervals[-1].tf
+        ):
             eom_intervals[-1].tf = total_duration
         # sampling the channel schedule
-        samples = sch.get_samples()
-        extended_samples = samples.extend_duration(total_duration)
-        for slot in sch:
-            if slot.ti == -1:
-                target["initial"] = slot.targets
-                continue
-            else:
-                # If slot is not the first element in schedule
-                if sch.in_eom_mode(slot):
-                    # EOM mode starts
-                    if not in_eom_mode:
-                        in_eom_mode = True
-                        eom_block_n += 1
-                elif in_eom_mode:
-                    # Buffer when EOM mode is disabled and next slot has 0 amp
-                    in_eom_mode = False
-                    if extended_samples.amp[slot.ti] == 0:
-                        eom_end_buffers[eom_block_n] = EOMSegment(
-                            slot.ti, slot.tf
-                        )
-                if (
-                    eom_block_n + 1 < nb_eom_intervals
-                    and slot.tf == eom_intervals[eom_block_n + 1].ti
-                    and extended_samples.det[slot.tf - 1]
-                    == sch.eom_blocks[eom_block_n + 1].detuning_off
-                ):
-                    # Buffer if next is eom and final det matches det_off
-                    eom_start_buffers[eom_block_n + 1] = EOMSegment(
-                        slot.ti, slot.tf
-                    )
+        extended_samples = ch_samples.extend_duration(total_duration)
+
+        for slot in ch_samples.slots:
+            # If slot is not the first element in schedule
+            if ch_samples.in_eom_mode(slot):
+                # EOM mode starts
+                if not in_eom_mode:
+                    in_eom_mode = True
+                    eom_block_n += 1
+            elif in_eom_mode:
+                # Buffer when EOM mode is disabled and next slot has 0 amp
+                in_eom_mode = False
+
+                tf = eom_intervals[eom_block_n].tf
+                if tf is not None and slot.ti != eom_intervals[eom_block_n].tf:
+                    if np.isclose(
+                        extended_samples.amp[
+                            tf
+                            + 2
+                            * cast(
+                                BaseEOM, sampled_seq._ch_objs[ch].eom_config
+                            ).rise_time
+                        ],
+                        0,
+                    ):
+                        eom_end_buffers[eom_block_n] = EOMSegment(tf, slot.ti)
```

Comment:
So, you are assuming that the entire time between the end of the EOM block and the next pulse is an end buffer, right? If one adds a delay after disabling EOM mode, then this delay will be included in the end buffer, right?
I think it's quite hard (if not impossible) to be accurately inferring the EOM end buffer like this. Personally, I'd be in favor of adding extra metadata to the `ChannelSamples` at the moment they are created to include this information. The `EOMSegment`s could even come already defined.

Reply:
Ok, I will do it this afternoon

Reply:
@a-corni `eom_end_buffers` and `eom_start_buffers` are list of EOMSegments, but essentially they are list of pairs (ti,tf). Can you provide me further details about what you are thinking?

Reply:
@dakk Indeed, I think the information of when a segment starts and ends should do, so storing the start and buffers as separate lists of tuples inside `ChannelSamples` should be enough. Then, when drawing, you can extract this information to recreate the `eom_start_buffers` and `eom_end_buffers` as lists of `EOMSegment`s